### DISCUSSION TITLE
How can I block get the dashboard list for public user?

### DISCUSSION THREAD

**Discussion by senturkayhan** [apache/superset#15838] [Q&A / Help]
<img width="1235" alt="Ekran Resmi 2021-07-22 16 32 51" src="https://user-images.githubusercontent.com/15243054/126647522-bf1c9299-8f2f-4a48-95fd-c68ba56bf8ce.png">

I had to give 'can read on Chart' and 'can read on  Dashboard' permissions to the public user for doesn't take 'unexpected error' on chart in dashboard. But now, the public user can reach the dashboard list and the chart list and then export the dashboard so the public user can read my query on the export file. I don't want the public user to can export the dashboard or reach the dashboard list. How can I block?

**Comment by senturkayhan** [comment]
I solved it

**Reply by sfirke** [reply]
@senturkayhan how did you solve it?  I'm interested in this.

**Reply by senturkayhan** [reply]
@sfirke If the public user reach the dashboard list, there is no any active user , so I controlled it in the dashboard views.py and then the public user is redirected to login page. 
I hope it can be help to you.


class DashboardModelView(
    DashboardMixin, SupersetModelView, DeleteMixin
):  # pylint: disable=too-many-ancestors
    route_base = "/dashboard"
    datamodel = SQLAInterface(DashboardModel)
    # TODO disable api_read and api_delete (used by cypress)
    # once we move to ChartRestModelApi
    class_permission_name = "Dashboard"
    method_permission_name = MODEL_VIEW_RW_METHOD_PERMISSION_MAP

    include_route_methods = RouteMethod.CRUD_SET | {
        RouteMethod.API_READ,
        RouteMethod.API_DELETE,
        "download_dashboards",
    }

    @has_access
    @expose("/list/")
    def list(self) -> FlaskResponse:
        print("User Control ---------------")
        print(g.user)
        print(g.user.is_active)
        if not g.user.is_active: 
            return redirect("/login")

     if not is_feature_enabled("ENABLE_REACT_CRUD_VIEWS"):
            return super().list()

        return super().render_app_template()

**Comment by sfirke** [comment]
Interesting, thanks for sharing this fix!  I wonder if this could be added
to base Superset, to block this page if not logged in.

**Reply by sfirke** [reply]
To summarize it looks like we could just add your two lines:
```
    if not g.user.is_active: 
        return redirect("/login")
```
ahead of this line: https://github.com/apache/superset/blob/44654e5abcc70d1fd934529dc82d7fe1d1bb20ef/superset/views/dashboard/views.py#L63

And it would do it.

**Reply by sfirke** [reply]
We should do the same to block listing of charts: https://github.com/apache/superset/blob/44654e5abcc70d1fd934529dc82d7fe1d1bb20ef/superset/views/chart/views.py#L61

**Reply by senturkayhan** [reply]
You should be carefull for the chart operation, If I'm right remember, the class which is include redirect function is not import the chart/views.py, you have to add this "from flask import g, redirect" at the top of this file.
